WalletWasabi/Blockchain/BlockFilters/IndexBuilderService.cs

```diff
@@ -118,148 +90,105 @@ public void Synchronize()
+										currentHeight = lastIndex.Header.Height;
 										currentHash = lastIndex.Header.BlockHash;
 									}
+									else
+									{
+										currentHash = StartingHeight == 0 ? uint256.Zero :
+											await RpcClient.GetBlockHashAsync((int)StartingHeight - 1);
+										currentHeight = StartingHeight - 1;
+									}
 								}
 
-								// If not synchronized or already 5 min passed since last update, get the latest blockchain info.
-								if (!syncInfo.IsCoreSynchornized || DateTimeOffset.UtcNow - syncInfo.BlockchainInfoUpdated > TimeSpan.FromMinutes(5))
+								var coreNotSynced = !syncInfo.IsCoreSynchornized;
+								var tipReached = syncInfo.BlockCount == currentHeight;
+								var refresh = DateTimeOffset.UtcNow - syncInfo.BlockchainInfoUpdated > TimeSpan.FromMinutes(5);
+								if (coreNotSynced || tipReached || refresh)
 								{
 									syncInfo = await GetSyncInfoAsync();
 								}
 
-								if (syncInfo.BlockCount - heightToRequest <= 100)
+								// If wasabi filter height is the same as core we may be done.
+								if (syncInfo.BlockCount == currentHeight)
 								{
-									// Both Wasabi and our Core node is in sync. Start doing stuff through P2P from now on.
-									if (syncInfo.IsCoreSynchornized && syncInfo.BlockCount == heightToRequest - 1)
+									// Check that core is fully synced
+									if (syncInfo.IsCoreSynchornized && !syncInfo.InitialBlockDownload)
 									{
-										syncInfo = await GetSyncInfoAsync();
-										// Double it to make sure not to accidentally miss any notification.
-										if (syncInfo.IsCoreSynchornized && syncInfo.BlockCount == heightToRequest - 1)
-										{
-											// Mark the process notstarted, so it can be started again and finally block can mark it is stopped.
-											Interlocked.Exchange(ref _running, 0);
-											return;
-										}
-									}
-
-									// Mark the synchronizing process is working with immature blocks from now on.
-									isImmature = true;
-								}
-
-								Block block = await RpcClient.GetBlockAsync(heightToRequest);
-
-								// Reorg check, except if we're requesting the starting height, because then the "currentHash" wouldn't exist.
-								if (heightToRequest != StartingHeight && currentHash != block.Header.HashPrevBlock)
-								{
-									// Reorg can happen only when immature. (If it'd not be immature, that'd be a huge issue.)
-									if (isImmature)
-									{
-										await ReorgOneAsync();
+										// Mark the process notstarted, so it can be started again
+										// and finally block can mark it as stopped.
+										Interlocked.Exchange(ref _serviceStatus, NotStarted);
+										return;
 									}
 									else
 									{
-										Logger.LogCritical("This is something serious! Over 100 block reorg is noticed! We cannot handle that!");
+										// Core is catching up give it a few milliseconds
+										await Task.Delay(100);
```

Comment:
100 milliseconds would imply 10 rpc call per second when we reach the tip while knots is catching up, it could be too much imo.

Reply:
hmmm good point, perhaps 10000ml is better